### PR TITLE
Throw an exception if configurationType to CreateConfigureDelegate is…

### PR DESCRIPTION
**Pull request by jbagga**
… abstract/has no ctor
Addresses #5431 
cc @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs [hunk 1]
```diff
@@ -23,6 +23,12 @@ public Action<IApplicationBuilder> CreateConfigureDelegate(Type configurationTyp
                 throw new ArgumentNullException(nameof(configurationType));
             }
 
+            if (!HasDefaultConstructor(configurationType.GetTypeInfo()))
+            {
+                throw new MissingMethodException(
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs]: `InvalidOperationException`

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs [hunk 2]
```diff
@@ -23,6 +23,12 @@ public Action<IApplicationBuilder> CreateConfigureDelegate(Type configurationTyp
                 throw new ArgumentNullException(nameof(configurationType));
             }
 
+            if (!HasDefaultConstructor(configurationType.GetTypeInfo()))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs]: HasParameterlessConstructor

File: src/Microsoft.AspNetCore.Mvc.Core/Resources.resx [hunk 3]
```diff
@@ -383,4 +383,7 @@
     <value>A duplicate entry for library reference {0} was found. Please check that all package references in all projects use the same casing for the same package references.</value>
     <comment>{0} is the dependency name.</comment>
   </data>
+  <data name="MiddlewareFilterConfigurationProvider_CreateConfigureDelegate_CannotCreateType" xml:space="preserve">
+    <value>Unable to create type {0}. The class is either abstract or no constructor was found.</value>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Resources.resx]: Unable to create an instance of type '{0}'. The type specified in `{1}` must not be abstract and must have a parameterless constructor.

1 -> nameof(ConfigureType)

**Comment by pranavkm** [comment]
cc @kichalla could you have a look?

**Comment by jbagga** [comment]
@pranavkm @kichalla Updated

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs [hunk 4]
```diff
@@ -14,6 +14,19 @@ namespace Microsoft.AspNetCore.Mvc.Internal
     public class MiddlewareFilterConfigurationProviderTest
     {
         [Fact]
+        public void CreateConfigureDelegate_ThrowsIfNoConstructorFoundForType()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs]: `CreateConfigureDelegate_ThrowsIfTypeIsAbstract`

Need another one where the type is non-abstract but does not have a parameterless ctor.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs [hunk 5]
```diff
@@ -66,6 +72,11 @@ private static MethodInfo FindMethod(Type startupType, Type returnType = null)
             return methodInfo;
         }
 
+        private bool HasParameterlessConstructor(TypeInfo modelTypeInfo)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterConfigurationProvider.cs]: `static`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs [hunk 6]
```diff
@@ -174,5 +187,9 @@ private void Configure(IApplicationBuilder appBuilder)
 
             }
         }
+
+        private abstract class AbstractType_NoConstructor
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs]: `AbstractType` should suffice. The presence or absence of ctor shouldn't matter.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs (hunk 4 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs]: Also try specifying a scenario with an interface. You could make this test use a `Theory` to test these scenarios.

**Comment by jbagga** [comment]
@pranavkm Updated

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs [hunk 7]
```diff
@@ -174,5 +190,16 @@ private void Configure(IApplicationBuilder appBuilder)
...
             }
         }
+
+        private abstract class AbstractType
+        {
+        }
+
+        private class NoParameterlessConstructor
+        {
+            private NoParameterlessConstructor(object A)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MiddlewareFilterConfigurationProviderTest.cs]: `public NoParameterlessConstructor(object a)`

**Comment by pranavkm** [comment]
:shipit: once @kichalla is happy

**Pull request merged**